### PR TITLE
Basic validate server endpoint. DSDEEPB-651

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  import context.dispatcher
+  import ValidateActor.ValidateWorkflow
+  private val tag = "ValidateActor"
+
+  override def receive = {
+    case ValidateWorkflow =>
+      val sentBy = sender()
+      logger.info(s"$tag for $sentBy")
+      val futureInputs = for {
+        eventualNamespace <- Future(NamespaceWithWorkflow.load(wdlSource, WorkflowManagerActor.BackendType))
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: Why is this namespace "eventual"?

File: src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 2]
```diff
@@ -170,6 +171,37 @@ trait CromwellApiService extends HttpService with PerRequestCreator {
       }
     }
 
+  @Path("/{version}/validate")
+  @ApiOperation(
+    value = "Validate a new workflow for execution",
```
Comment (mcovarr) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: IMHO "new" is kind of confusing here

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+
+class ValidateActor(wdlSource: WdlSource, wdlJson: WdlJson)
+  extends Actor with LazyLogging {
+
+  import WorkflowJsonSupport._
+  import context.dispatcher
+  import ValidateActor.ValidateWorkflow
+  private val tag = "ValidateActor"
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: If this is literally a static constant should it be in the companion object?

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  override def receive = {
+    case ValidateWorkflow =>
+      val sentBy = sender()
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: could we put this block in a function which gets called by the case?

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      logger.info(s"$tag for $sentBy")
+      val futureInputs = for {
+        eventualNamespace <- Future(NamespaceWithWorkflow.load(wdlSource, WorkflowManagerActor.BackendType))
+        inputs <- Future(wdlJson.parseJson) map {_.asJsObject.fields}
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: as much as I dislike it, I was skooled recently that you're supposed to have spaces in side the curly braces on those HOF closures

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        coercedInputs <- Future.fromTry(eventualNamespace.coerceRawInputs(inputs))
+        _ <- Future.fromTry(eventualNamespace.staticDeclarationsRecursive(coercedInputs))
+      } yield ()
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: While I run the risk of looking like a dope on the internet here, what is actually being assigned to `futureInputs` if it's yielding `()`?

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+      futureInputs onComplete {
+        case Success(_) =>
+          logger.info(s"$tag success $sentBy")
+          sentBy ! RequestComplete(StatusCodes.OK, WorkflowValidateResponse(valid = true, error = None))
+
+        case Failure(ex) =>
+          val messageOrBlank = Option(ex.getMessage).mkString
+          logger.info(s"$tag error $sentBy: $messageOrBlank")
+          sentBy ! RequestComplete(StatusCodes.BadRequest, WorkflowValidateResponse(valid = false, error = Option(messageOrBlank)))
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: Why is `messageOrBlank` option-i-fied above, turned into a string and optionated again?

File: src/main/scala/cromwell/webservice/ApiDataModels.scala [hunk 8]
```diff
@@ -7,54 +7,67 @@ import cromwell.engine.backend.StdoutStderr
 
 import scala.annotation.meta.field
 
+@ApiModel(value = "WorkflowValidate")
+case class WorkflowValidateResponse
+(
+  @(ApiModelProperty@field)(required = true, value = "The validation of the workflow")
```
Comment (geoffjentry) [inline src/main/scala/cromwell/webservice/ApiDataModels.scala]: Is this swagger stuff? If 'yes' I hate it even more than I thought. If 'no' why is this here?

File: src/main/scala/cromwell/webservice/CromwellApiService.scala (hunk 2 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/webservice/CromwellApiService.scala]: :+1:

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala (hunk 1 above)
Comment (kshakir) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: Copypasta (violating DRY principles) from [`WorkflowManagerActor`](https://github.com/broadinstitute/cromwell/blob/f347e80b715b0959c52f5ce0abda09c549058a72/src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala#L168). Renamed to `namespaceWithWorkflow` to match the class name.

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala (hunk 5 above)
Comment (kshakir) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: Reverted to `.map()` as intellij wanted to line wrap the curly braces with spaces.

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala (hunk 6 above)
Comment (kshakir) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: Updated with a comment, and fixed the changed `futureInputs` to `futureValidation`… as there really weren't any inputs being yielded.

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala (hunk 7 above)
Comment (kshakir) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: I played a bit with replacing this with `val messageOrBlank = if (ex.getMessage == null) "" else ex.getMessage`, but the current line was more compact, and didn't use the "evil" `null`.

As to why when `ex.getMessage == null` the error is reported as with `Some("")` instead of `None`, I preferred letting the caller know that there _was_ an exception, but that the message wasn't there.

We could instead return a json `"error": null` instead of `"error": ""`, or completely eliminate the key altogether like we do when return `"valid": true`, but my preference was:

```json
{
  "valid": false,
  "error": ""
}
```

This also matches the behavior in [`WorkflowManagerActor`](https://github.com/broadinstitute/cromwell/blob/f347e80b715b0959c52f5ce0abda09c549058a72/src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala#L184)

File: src/main/scala/cromwell/webservice/ApiDataModels.scala (hunk 8 above)
Comment (kshakir) [inline src/main/scala/cromwell/webservice/ApiDataModels.scala]: This is a combo of stuff:

`@ApiModelProperty` is straight from the dark heart of [`swagger-core`](https://github.com/swagger-api/swagger-core/blob/v1.3.12/modules/swagger-annotations/src/main/java/com/wordnik/swagger/annotations/ApiModelProperty.java#L27).

`@field` is the extra special scala [magic](http://www.scala-lang.org/api/2.11.7/index.html#scala.annotation.meta.package).

The lack of spaces is the nice touch of IntelliJ IDEA's scala plugin + default auto formatting.

Comment (geoffjentry) [inline src/main/scala/cromwell/webservice/ApiDataModels.scala]: "The lack of spaces is the nice touch of IntelliJ IDEA's scala plugin" this made me chuckle

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 9]
```diff
@@ -244,6 +245,85 @@ class CromwellApiServiceSpec extends FlatSpec with CromwellApiService with Scala
       }
   }
 
+  s"Cromwell validate workflow API $version" should "return 200 for a successful workflow submission" in {
```
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: I think this means 'validation' not 'submission'?
